crates/services/gas_price_service/src/lib.rs

```diff
@@ -0,0 +1,234 @@
+    A: Send + Sync,
+{
+    const NAME: &'static str = "GasPriceUpdater";
+    type SharedData = SharedGasPriceAlgo<A>;
+    type Task = Self;
+    type TaskParams = ();
+
+    fn shared_data(&self) -> Self::SharedData {
+        self.next_block_algorithm.clone()
+    }
+
+    async fn into_task(
+        self,
+        _state_watcher: &StateWatcher,
+        _params: Self::TaskParams,
+    ) -> anyhow::Result<Self::Task> {
+        Ok(self)
+    }
+}
+
+#[async_trait]
+impl<A, U> RunnableTask for GasPriceService<A, U>
+where
+    U: UpdateAlgorithm<Algorithm = A> + Send + Sync,
+    A: Send + Sync,
+{
+    async fn run(&mut self, watcher: &mut StateWatcher) -> anyhow::Result<bool> {
+        let should_continue;
+        tokio::select! {
+            biased;
+            _ = watcher.while_started() => {
+                tracing::debug!("Stopping gas price service");
+                should_continue = false;
+            }
+            new_algo = self.update_algorithm.next() => {
+                tracing::debug!("Updating gas price algorithm");
+                self.update(new_algo).await;
+                should_continue = true;
+            }
+        }
+        Ok(should_continue)
+    }
+
+    async fn shutdown(self) -> anyhow::Result<()> {
+        Ok(())
+    }
+}
+
+#[allow(non_snake_case)]
+#[cfg(test)]
+mod tests {
+    use crate::{
+        GasPriceService,
+        UpdateAlgorithm,
+    };
+    use fuel_core_services::{
+        RunnableService,
+        Service,
+        ServiceRunner,
+        State,
+        StateWatcher,
+    };
+    use fuel_core_types::fuel_types::BlockHeight;
+    use tokio::sync::mpsc;
+
+    #[derive(Clone, Debug)]
+    struct TestAlgorithm {
+        price: u64,
+    }
+
+    impl TestAlgorithm {
+        fn gas_price(&self) -> u64 {
+            self.price
+        }
+    }
+
+    struct TestAlgorithmUpdater {
+        start: TestAlgorithm,
+        price_source: mpsc::Receiver<u64>,
+    }
+
+    #[async_trait::async_trait]
+    impl UpdateAlgorithm for TestAlgorithmUpdater {
+        type Algorithm = TestAlgorithm;
+
+        fn start(&self, _for_block: BlockHeight) -> Self::Algorithm {
+            self.start.clone()
+        }
+
+        async fn next(&mut self) -> Self::Algorithm {
+            let price = self.price_source.recv().await.unwrap();
+            TestAlgorithm { price }
+        }
+    }
+    #[tokio::test]
+    async fn run__updates_gas_price() {
+        let _ = tracing_subscriber::fmt::try_init();
+
+        // given
+        let (price_sender, price_receiver) = mpsc::channel(1);
+        let updater = TestAlgorithmUpdater {
+            start: TestAlgorithm { price: 0 },
+            price_source: price_receiver,
+        };
+        let service = GasPriceService::new(0.into(), updater);
+        let (watch_sender, watch_receiver) = tokio::sync::watch::channel(State::Started);
+        let watcher = StateWatcher::from(watch_receiver);
```

Comment:
You can re-use helper function 
<img width="877" alt="image" src="https://github.com/FuelLabs/fuel-core/assets/18346821/d2c22ede-dad4-433e-aaea-1e6cacd13b5a">